saleor/graphql/core/mutations.py

```diff
@@ -65,146 +47,159 @@ def _update_mutation_arguments_and_fields(cls, arguments, fields):
         cls._meta.fields.update(fields)
 
 
-class ModelFormMutationOptions(MutationOptions):
-    form_class = None
-    return_field_name = None
-
-
-class ModelDeleteMutationOptions(MutationOptions):
-    model = None
-    return_field_name = None
-
-
-class ModelDeleteMutation(BaseMutation):
-
+class ModelMutation(BaseMutation):
     class Meta:
         abstract = True
 
     @classmethod
     def __init_subclass_with_meta__(
-            cls, arguments=None, model=None, return_field_name=None,
-            _meta=None, **options):
+            cls, arguments=None, model=None, permissions=None, _meta=None,
+            **options):
         if not model:
-            raise ImproperlyConfigured(
-                'model is required for ModelDeleteMutation')
-
-        _meta = ModelDeleteMutationOptions(cls)
-
-        if not return_field_name:
-            return_field_name = get_model_name(model)
+            raise ImproperlyConfigured('model is required for ModelMutation')
+        _meta = ModelMutationOptions(cls)
+        return_field_name = get_model_name(model)
         if arguments is None:
             arguments = {}
-        arguments.update({'id': graphene.ID()})
         fields = get_output_fields(model, return_field_name)
-
         _meta.model = model
         _meta.return_field_name = return_field_name
-
         super().__init_subclass_with_meta__(_meta=_meta, **options)
-
-        # Update mutation's arguments and fields
         cls._update_mutation_arguments_and_fields(
             arguments=arguments, fields=fields)
 
     @classmethod
-    def mutate(cls, root, info, id, **kwargs):
-        model = cls._meta.model
-        model_type = registry.get_type_for_model(model)
-        instance = get_node(info, id, only_type=model_type)
-        instance.delete()
-        field_name = cls._meta.return_field_name
-        kwargs = {field_name: instance}
-        return cls(**kwargs)
+    def add_error(cls, errors, field, message):
+        errors.append(Error(field=field, message=message))
+        return errors
 
+    @classmethod
+    def _check_type(cls, field, target_type):
+        if hasattr(field.type, 'of_type'):
+            return field.type.of_type == target_type
+        return field.type == target_type
 
-class ModelFormMutation(BaseMutation):
-
-    class Meta:
-        abstract = True
+    @classmethod
+    def _clean_input(cls, info, instance, input, errors):
+        InputCls = getattr(cls.Arguments, 'input')
+        cleaned_input = {}
+        for field_name, field in InputCls._meta.fields.items():
+            if field_name in input:
+                value = input[field_name]
+                if value is not None:
+                    # FIXME: maybe we could have custom input field type that takes
```

Comment:
This may be an inspiration
https://github.com/graphql-python/graphene-django/blob/master/graphene_django/forms/forms.py

Reply:
I'll take a look. I have some ideas how to do this but I want to do it separately, to not block this PR anymore.